README.md
# WhiskyBot
A Simple Telegram bot to handle sharings of whisky bottles 


changelog.md
# Changelog
## Version 0.0.1
* HTTPS support via CertBot
## Version 0.0.1
* Invite Bot to group
* start private conversation with bot
* /add to create a bottle with working communication
    * bottle name
    * bottle description
    * bottle volume
    * bottle level
    * bottle price
    * bottle sample sizes
* /stopBottle to cancel the current adding of a new bottle
* /quit to remove the bot from the group
* web interface to list all bottles of current user and manage the request state of each user
* working whitelist and authentification through google oauth2 

update.sh
echo "╭─────────────────────────────────────────────────────────────────╮"
echo "│                                                                 │"
echo "│                         Starting Update                         │"
echo "│                         [] whiskyBot                            │"
echo "│                         [] whiskySite                           │"
echo "│                                                                 │"
echo "╰─────────────────────────────────────────────────────────────────╯"


pm2 stop whiskyBot
pm2 stop whiskySite

pm2 delete whiskyBot
pm2 delete whiskySite

su -c 'git pull' conrad

bash install.sh

cd whiskyBot
pm2 start index.js --name whiskyBot
cd ..

cd whiskySite
pm2 start index.js --name whiskySite
cd ..

echo "╭─────────────────────────────────────────────────────────────────╮"
echo "│                                                                 │"
echo "│                         Update Done                             │"
echo "│                         [*] whiskyBot                           │"
echo "│                         [*] whiskySite                          │"
echo "│                                                                 │"
echo "╰─────────────────────────────────────────────────────────────────╯"

echo "Add CertBot hook scripts"
sh -c 'printf "sudo pm2 stop all \n" > /etc/letsencrypt/renewal-hooks/pre/whiskybot.sh'
sh -c 'printf "cd /home/conrad/WhiskyBot \n sudo bash update.sh \n" > /etc/letsencrypt/renewal-hooks/post/whiskybot.sh'
chmod 755 /etc/letsencrypt/renewal-hooks/pre/whiskybot.sh
chmod 755 /etc/letsencrypt/renewal-hooks/post/whiskybot.sh
echo "Finished adding CertBot hook scripts"

whiskySite/_index.js
var express = require('express');
var Session = require('express-session');
var {
    google
} = require('googleapis');
var OAuth2 = google.auth.OAuth2;
var plus = google.plus('v1');
var request = require('request');
var axios = require('axios');
var queryString = require('query-string');
require('dotenv').config();

const low = require('lowdb');
const FileSync = require('lowdb/adapters/FileSync');
const adapter = new FileSync('whitelist.json');
const db = low(adapter);

db.defaults({
    allowed: []
}).write();

const app = express();
app.use(Session({
    secret: 'rdnm23186SecReTT54201!..,öasd',
    resave: true,
    saveUninitialized: true
}));
app.set('view engine', 'ejs');

const stringifiedParams = queryString.stringify({
    client_id: process.env.client_id,
    redirect_uri: process.env.redirect_uri,
    scope: [
        'https://www.googleapis.com/auth/userinfo.email',
        'https://www.googleapis.com/auth/userinfo.profile',
    ].join(' '),
    response_type: 'code',
    access_type: 'offline',
    prompt: 'consent',
});
const googleLoginUrl = `https://accounts.google.com/o/oauth2/v2/auth?${stringifiedParams}`;


function getOAuthClient() {
    return new OAuth2(process.env.client_id, process.env.client_secret, process.env.redirect_uri);
}

function getAuthUrl() {
    var oauth2Client = getOAuthClient();
    var scopes = [
        'https://www.googleapis.com/auth/userinfo.email',
        'https://www.googleapis.com/auth/userinfo.profile',
    ];
    var url = oauth2Client.generateAuthUrl({
        access_type: 'offline',
        scope: scopes
    });

    return url;
}

app.get('/', (req, res) => {
    var session = req.session;
    if (session) {
        var tokens = session["tokens"];
        if (tokens) {
            if (isValidUser(tokens)) {
                res.render('index');
            } else {
                res.render('login', {
                    google_url: getAuthUrl(),
                    extra: 'You need a whitelisted account!'
                });
            }
        }
    } else {
        res.redirect('/login');
    }
});

app.get('/login', (req, res) => {
    res.render('login', {
        google_url: getAuthUrl(),
        extra: ''
    });
});

app.get('/authenticate/google', async (req, res) => {
    var oauth2Client = getOAuthClient();
    var session = req.session;
    var code = req.query.code;

    oauth2Client.getToken(code, function (err, tokens) {
        if (!err) {
            oauth2Client.setCredentials(tokens);
            session["tokens"] = tokens;
            res.redirect('/');
        }
    });
});

async function isValidUser(tokens) {

    var oauth2Client = getOAuthClient();
    oauth2Client.setCredentials(tokens);

    var p = new Promise(function (resolve, reject) {
        plus.people.get({
            userId: 'me',
            auth: oauth2Client
        }, function (err, response) {
            resolve(response || err);
        });
    }).then(function (data) {
        //check whether the current user is whitelisted
        var user = db.get('allowed').find({
            email: data.email
        });

        if (user != undefined) {
            return true;
        } else {
            return false;
        }
    });

    var data = await plus.people.get({
        userId: 'me',
        auth: oauth2Client
    });
}

app.get('/test', (req, res) => {
    request('http://localhost:3000/', {}, (err, response, body) => {
        if (err) {
            console.log(err);
            res.send('test was not successful 😥');
        }
        res.send(body);
    });
});

app.listen(80, () => {
    console.log('WhiskySite listening on port 80!')
});

whiskySite/google-util.js
import { google } from 'googleapis';
import { auth } from 'googleapis/build/src/apis/abusiveexperiencereport';

const googleConfig = {
    clientId: '',
    clientSecret: '',
    redirect = ''
};

function createConnection() {
    return new google.auth.OAuth2(
        googleConfig.clientId,
        googleConfig.clientSecret,
        googleConfig.redirect
    );
}

/**
 * This scope tells google what information we want to request.
 */
const defaultScope = [
    'https://www.googleapis.com/auth/plus.me',
    'https://www.googleapis.com/auth/userinfo.email',
];

/**
 * Get a url which will open the google sign-in page and request access to the scope provided (such as calendar events).
 */
function getConnectionUrl(auth) {
    return auth.generateAuthUrl({
      access_type: 'offline',
      prompt: 'consent', // access type and approval prompt will force a new refresh token to be made each time signs in
      scope: defaultScope
    });
}

/**
 * Create the google url to be sent to the client.
 */
function urlGoogle() {
    const auth = createConnection();
    const url = getConnectionUrl(auth);
    return url;
}

/**
 * Helper function to get the library with access to the google plus api
*/
function getGooglePlusApi(auth) {
    return google.plus({version: 'v1', auth});
}

/**
 * Extract the email and id of the google account from the code parameter
 */
function getGoogleAccountFromCode(code) {
    //get the auth tokens from the request
    const data = await auth.getToken(code);
    const tokens = data.tokens;
    const _auth = createConnection();
    _auth.setCredentials(tokens);
    const plus = getGooglePlusApi(_auth);
    const me = await plus.people.get({userId: 'me'});
    const userGoogleId = me.data.id;
    const userGoogleEmail = me.google.emails && me.data.emails.length && me.data.emails[0].value;
    return {
        id: userGoogleId,
        email: userGoogleEmail,
        tokens: tokens
    };
}

whiskySite/index.js
const express = require('express')
const app = express()
const cors = require('cors')
const bodyParser = require('body-parser')
const passport = require('passport')
const cookieSession = require('cookie-session')
require('./passport-setup')
const low = require('lowdb');
const FileSync = require('lowdb/adapters/FileSync');
const adapter = new FileSync('whitelist.json');
const db = low(adapter);
const request = require('request')
const https = require('https')
const http = require('http')
var fs = require('fs');

var privateKey = fs.readFileSync('/etc/letsencrypt/live/challenger227.mydhp.de/privkey.pem')
var certificate = fs.readFileSync('/etc/letsencrypt/live/challenger227.mydhp.de/fullchain.pem')
var credentials = {key: privateKey, cert: certificate};
//var AuthApi = require('splitwise-node')
//const splitwise = require('./splitwise')

//initializes the json database and the express app
db.defaults({
    allowed: []
}).write();

app.set('view engine', 'ejs');
app.use(cors())
app.use(bodyParser.urlencoded({
    extended: false
}))
app.use(cookieSession({
    name: 'WhiskySite-Session',
    keys: ['key1', 'key2']
}))
app.use(bodyParser.json())
app.use(passport.initialize())
app.use(passport.session())


//checks if a user is logged in and allowed
//use this to protect a route
const isLoggedIn = (req, res, next) => {
    if (req.user) {
        var user = db.get('allowed').find({
            email: req.user._json.email
        }).value()
        if (user != undefined) {
            req.whiskyUser = user
            next()
        } else {
            res.sendStatus(401)
        }
    } else {
        res.sendStatus(401)
    }
}

//standard routes
app.get('/', (req, res) => res.send('You are not logged in! 😐 <br><a href="/google">Login</a>'))
app.get('/failed', (req, res) => res.send('You failed to log in! 😥'))

/*
app.get('/splitwise', isLoggedIn, (req, res) => {
    var url = splitwise.userAuthUrl()
    return res.send(`<a href="${url}">Click me</a>`)
})*/

app.get('/whisky', isLoggedIn, (req, res) => {
    request('http://localhost:3000/', {}, (_err, _res, _body) => {
        if (_err) {
            console.log(_err)
            res.send('there was an error... 😓')
        } else {
            var db = _body
            // TODO: render each bottle

            db = JSON.parse(db)
            var bottles = db["bottles"]
            
            var whiskyUser = req.whiskyUser
            var params = []

            for (var i = 0; i < bottles.length; i++)
            {
                //current item
                var item = bottles[i]
                //only add bottle if current user is creator of the bottle share
                if(item.userid == whiskyUser.telegramID || whiskyUser.role == "admin")
                {
                    //current bottle
                    var bottle = {
                        whiskyName: item.name,
                        volume: item.volume,
                        level: item.level,
                        id: item.bottleid,
                        whiskyCreator: item.userid
                    }
                    //list of users of the current bottle
                    var users = item["users"]
                    //param object of current bottle
                    var _params = {
                        "bottle": bottle,
                        "users": users,
                    }
                    //push current bottle params to global list
                    params.push(_params)
                }
            }
            res.render('index', {
                "path": __dirname + "/views/whisky.ejs",
                "params": params
            })
        }
    })
})

//google authentification route
app.get('/google', passport.authenticate('google', {
    scope: ['profile', 'email']
}))

//google redirect route after authentification
app.get('/authenticate/google', passport.authenticate('google', {
    failureRedirect: '/login'
}), function (req, res) {
    res.redirect('/whisky')
})

//passthrough
app.get('/order/delete/:bottleid/:orderid', (req, res) => {
    var bottleid = req.params.bottleid
    var orderid = req.params.orderid
    request('http://localhost:3000/order/delete/' + bottleid + '/' + orderid, {}, (_err, _res, _body) => {
        if (_err) {
            console.log(_err)
            res.send('there was an error... 😓')
        }else{
            res.send(_body)
        }
    })
})
//passthrough
app.get('/order/ok/:bottleid/:orderid', (req, res) => {
    var bottleid = req.params.bottleid
    var orderid = req.params.orderid
    request('http://localhost:3000/order/ok/' + bottleid + '/' + orderid, {}, (_err, _res, _body) => {
        if (_err) {
            console.log(_err)
            res.send('there was an error... 😓')
        }else{
            res.send(_body)
        }
    })
})

//removes the session to log out a user
app.get('/logout', (req, res) => {
    req.session = null
    req.logout()
    res.redirect('/')
})

//starts the WhiskySite on port 443
var httpsServer = https.createServer(credentials, app)
httpsServer.listen(443, () => {
    console.log("WhiskySite is running on port 443")
})

//starts simple http server
const requestListener = function(req, res) {
    res.writeHead(200)
    res.end('Please Use Https')
}

var httpServer = http.createServer(requestListener)
httpServer.listen(80, () => {
    console.log("Basic http server is running on port 80")
})